File: rxauth/src/main/java/com/sebastianmarschall/rxsmartlock/RxAuth.java
package com.sebastianmarschall.rxsmartlock;

import android.content.Context;
import android.content.Intent;

import com.google.android.gms.auth.api.credentials.Credential;
import com.google.android.gms.auth.api.credentials.CredentialPickerConfig;
import com.google.android.gms.auth.api.credentials.CredentialRequest;
import com.sebastianmarschall.rxsmartlock.observable.RetrieveCredentialObservable;

import io.reactivex.Observable;
import io.reactivex.Single;

public class RxAuth {

    private Context context;
    private CredentialRequest credentialRequest;


    private RxAuth(Builder builder) {
        context = builder.getContext();
        credentialRequest = builder.getCredentialRequestBuilder().build();
    }

    public Single<Credential> retrieveCredentials() {
        return Single.create(new RetrieveCredentialObservable(context, credentialRequest));
    }

    public Single<Credential> retrieveCredentialFromIntent(Intent data) {
        return Single.just((Credential) data.getParcelableExtra(Credential.EXTRA_KEY));
    }

    public static class Builder {

        private Context mContext;
        private CredentialRequest.Builder mCredentialRequestBuilder;

        public Builder(Context context) {
            mContext = context;
            mCredentialRequestBuilder = new CredentialRequest.Builder();
        }

        public Builder setAccountTypes(String... accountTypes) {
            mCredentialRequestBuilder.setAccountTypes(accountTypes);
            return this;
        }

        public Builder setCredentialHintPickerConfig(CredentialPickerConfig config) {
            mCredentialRequestBuilder.setCredentialHintPickerConfig(config);
            return this;
        }

        public Builder setCredentialPickerConfig(CredentialPickerConfig config) {
            mCredentialRequestBuilder.setCredentialPickerConfig(config);
            return this;
        }

        public Builder setPasswordLoginSupported(boolean passwordLoginSupported) {
            mCredentialRequestBuilder.setPasswordLoginSupported(passwordLoginSupported);
            return this;
        }

        public RxAuth build() {
            return new RxAuth(this);
        }

        private Context getContext() {
            return mContext;
        }

        private CredentialRequest.Builder getCredentialRequestBuilder() {
            return mCredentialRequestBuilder;
        }

    }
}

File: rxauth/src/main/java/com/sebastianmarschall/rxsmartlock/RxAuth.kt
package com.sebastianmarschall.rxsmartlock

import android.content.Context
import android.content.Intent
import android.os.Parcelable
import com.google.android.gms.auth.api.credentials.Credential
import com.google.android.gms.auth.api.credentials.CredentialPickerConfig
import com.google.android.gms.auth.api.credentials.CredentialRequest
import com.sebastianmarschall.rxsmartlock.observable.DeleteCredentialObservable
import com.sebastianmarschall.rxsmartlock.observable.RetrieveCredentialObservable
import com.sebastianmarschall.rxsmartlock.observable.StoreCredentialObservable
import io.reactivex.Single

class RxAuth private constructor(builder: Builder) {

    private val context: Context
    private val credentialRequest: CredentialRequest

    init {
        context = builder.context
        credentialRequest = builder.credentialRequestBuilder.build()
    }

    fun retrieveCredentials(): Single<Credential> {
        return Single.create(RetrieveCredentialObservable(context, credentialRequest))
    }

    fun retrieveCredentialFromIntent(data: Intent): Single<Credential> {
        return Single.just(data.getParcelableExtra<Parcelable>(Credential.EXTRA_KEY) as Credential)
    }

    fun storeCredentials(credential: Credential): Single<Boolean> {
        return Single.create(StoreCredentialObservable(context, credential))
    }

    fun deleteCredential(credential: Credential): Single<Boolean> {
        return Single.create(DeleteCredentialObservable(context, credential))
    }

    class Builder(val context: Context) {
        val credentialRequestBuilder: CredentialRequest.Builder = CredentialRequest.Builder()

        fun setAccountTypes(vararg accountTypes: String): Builder {
            credentialRequestBuilder.setAccountTypes(*accountTypes)
            return this
        }

        fun setCredentialHintPickerConfig(config: CredentialPickerConfig): Builder {
            credentialRequestBuilder.setCredentialHintPickerConfig(config)
            return this
        }

        fun setCredentialPickerConfig(config: CredentialPickerConfig): Builder {
            credentialRequestBuilder.setCredentialPickerConfig(config)
            return this
        }

        fun setPasswordLoginSupported(passwordLoginSupported: Boolean): Builder {
            credentialRequestBuilder.setPasswordLoginSupported(passwordLoginSupported)
            return this
        }

        fun build(): RxAuth {
            return RxAuth(this)
        }

    }
}
